NestedFolders: Add folder service registry with dashboard service implementation

suntala opened the pull request:
**What is this feature?**
It introduces a registry of interfaces for Folder service by having the Dashboard service implement it. The registry has purposefully been kept minimal for now. Only a deletion method has been included.

**Why do we need this feature?**
To simplify the introduction of services which are "children" to the folder service and enable functionality such as counting and deleting all descendants of a given folder

**Who is this feature for?**
It supports the nested folders feature which is for the users referenced in the [Product DNA doc](https://docs.google.com/document/d/1xGkmfP5iemurkyAoHXx6BxeewurZumFkc9TTcUIYN34/edit#)

**Which issue(s) does this PR fix?**:
Partial https://github.com/grafana/grafana/issues/63872. Implementing counting will be done in a separate PR.

**Special notes for your reviewer**:
- Corresponding enterprise PR: https://github.com/grafana/grafana-enterprise/pull/4880.
- Size of the PR: The main reason it's large is because the [mockery file for the dashboard store](pkg/services/dashboards/store_mock.go) had to be regenerated.
- Future improvements: 
  - [ ] Generic sync mutex: it would be great to add it as a util to be used by the folder service as well as quota. I would prefer to do it as a separate PR because it would keep the changes introduced in this PR minimal and they would be fairly self-contained.
- Deletion in a transaction: there’s been discussion about atomic vs partial deletion (see [nested folder sync notes](https://docs.google.com/document/d/1qonh75vgI0dYPU6acuwqQnXZ9T_8MHsyo-X4ZO2SiR4/edit#) for March 22nd, 2023). I think it’s still worth looking into supporting partial deletion. However, in the interest of keeping the changes in the PR minimal, I would suggest exploring this in a separate PR if we do decide to pursue this option. This is because removing the transaction around the deletion of the folder would open up a new set of issues. For instance, it could lead to a folder being deleted in the folder table (takes place due to [nestedFolderDelete()](https://github.com/grafana/grafana/blob/suntala/reg-service/pkg/services/folder/folderimpl/folder.go#L448) call —> [here](https://github.com/grafana/grafana/blob/suntala/reg-service/pkg/services/folder/folderimpl/folder.go#L612)), but not its related permission (takes place due to [legacyDelete()](https://github.com/grafana/grafana/blob/suntala/reg-service/pkg/services/folder/folderimpl/folder.go#L479) call —> [here](https://github.com/grafana/grafana/blob/dc147f18d725d5b2d661713313a78b0e2b9c8670/pkg/services/dashboards/database/database.go#L721)).
- Deletions related to dashboard deletions: there's more being done in the `DeleteDashboardsInFolder` method than just the deletion of the dashboards in a given folder(see [deleteChildrenDashboardAssociations](https://github.com/grafana/grafana/pull/65033/files#diff-c0f0aa2cad053ea6fc9be737cc6ad6d4145ceb39578ec3710d7eea2442fd2757R1067)). When we delete a dashboard, we also need to remove table entries which reference it (for example in dashboard_tag, star etc). Referring back to the original issue, these are reference relations which is why I didn't pull them out into their own registry implementations. The goal outlined in the original ticket was:
>  instead of solving two problems with a single solution, let's focus on Association problem first and limit its scope to the Folder service specifically. Let's keep Reference manager/registry for later.
- Benchmarking: some smaller tests indicate that the deletion time for the new deletion method via registry is comparable (2m9s vs 2m11s in this case) to the existing deletion time. This is the [k6 test](https://github.com/grafana/grafana-api-tests/compare/suntala/nested-folder-deletion?expand=1) that was run. Check the top of each screenshot to see how many dashboards/folders were involved. 
<details>
  <summary>Screenshots</summary>
Old
<img width="801" alt="Comparable time!! Screenshot 2023-04-12 at 10 56 22" src="https://user-images.githubusercontent.com/33031346/231451965-3e1cf5f0-cadf-4403-b2fc-13ddf4b08062.png">
New
<img width="845" alt="Screenshot 2023-04-12 at 10 46 01" src="https://user-images.githubusercontent.com/33031346/231452176-2795eada-1e7e-49aa-986b-4a4cdce42d73.png">
</details>

suntala committed:
Delete folders, dashboards with registry service

suntala committed:
Remove folder deletion transactions

suntala committed:
Uncomment integration test

suntala committed:
Update signature of ProvideDashboardServiceImpl

suntala committed:
Remove separate registry package

suntala committed:
Rename folder registry service

suntala committed:
Update target service conflict error

suntala committed:
Regenerate mockery file

suntala committed:
Add test for DeleteInFolder

suntala committed:
Add test for DeleteDashboardsInFolder

suntala committed:
Remove unnecessary deleteChildrenInFolder call
since it is already being called in the main Delete method

suntala committed:
Delete child dashboard associations via registry

suntala committed:
Add back folder deletion transactions

suntala committed:
Refactor deleteChildrenDashboardAssociations

suntala committed:
Improve DeleteDashboardsInFolder test

sakjur commented on pkg/services/dashboards/database/database.go:
```diff
@@ -809,6 +753,74 @@ func (d *dashboardStore) deleteDashboard(cmd *dashboards.DeleteDashboardCommand,
+	return nil
+}
+
+func deleteFolderRules(sess *db.Session, dashboard dashboards.Dashboard, forceDeleteFolderRules bool) error {
```
```suggestion
func deleteFolderAlertRules(sess *db.Session, dashboard dashboards.Dashboard, forceDeleteFolderRules bool) error {
```

Could we rename this function to be a little bit precise? It took me a little bit of time to realize this wasn't about permissions 🙂

sakjur commented on pkg/services/folder/model.go:
```diff
@@ -13,6 +13,7 @@ var ErrDatabaseError = errutil.NewBase(errutil.StatusInternal, "folder.database-
 var ErrInternal = errutil.NewBase(errutil.StatusInternal, "folder.internal")
 var ErrFolderTooDeep = errutil.NewBase(errutil.StatusInternal, "folder.too-deep")
 var ErrCircularReference = errutil.NewBase(errutil.StatusBadRequest, "folder.circular-reference", errutil.WithPublicMessage("Circular reference detected"))
+var ErrTargetRegistrySrvConflict = errutil.NewBase(errutil.StatusBadRequest, "folder.target-registry-srv-conflict")
```
I think this should be an internal error, since the user isn't supposed to end up getting this error unless something is _very_ wrong?

suntala replied:
Yes, that makes sense. I had copied this over from [quota service](https://github.com/grafana/grafana/blob/main/pkg/services/quota/model.go#L17), but it's a different context there.

sakjur commented on pkg/services/folder/folderimpl/folder.go:
```diff
@@ -477,6 +487,15 @@ func (s *Service) Delete(ctx context.Context, cmd *folder.DeleteFolderCommand) e
 	return err
 }
 
+func (s *Service) deleteChildrenInFolder(ctx context.Context, orgID int64, UID string) error {
```
We could add checks here to ensure that we don't accidentally pass orgID 0 or the empty UID string to the services, which should reduce the risk of a small bug somewhere becoming a large bug.

sakjur reviewed (approved):
👍 Tested this and it works great!

I left some minor comments that you don't need to address in this PR, but you're welcome to do so if you'd like.

It seems to me like we're looking at deleting alert rules from the alerting service rather than from the dashboards service, which is why `DeleteDashboardsInFolder` doesn't call `deleteFolderRules`, is that correct?

suntala committed:
Rename deleteFolderRules to deleteFolderAlertRules

suntala committed:
Fix ErrTargetRegistrySrvConflict

suntala committed:
Add validation of folder uid and org id

suntala commented:
@sakjur thanks for the feedback! I addressed your comments in the last three commits.

> It seems to me like we're looking at deleting alert rules from the alerting service rather than from the dashboards service, which is why DeleteDashboardsInFolder doesn't call deleteFolderRules, is that correct?

The reason why `DeleteDashboardsInFolder` doesn't call `deleteFolderRules/deleteFolderAlertRules` is because  `DeleteDashboardsInFolder` was originally only meant to delete dashboards. However, in this function, I had to also remove table entries which reference them since this is the only place we have access to dashboard identifiers. I'd be interested to explore better options though.

My plan was to address the deletion of alerts in a separate PR. I've basically left the call to `deleteFolderRules/deleteFolderAlertRules` where it has always been. I've just isolated it into a separate function. I thought it made it easier to follow what was happening in `deleteDashboard` and because we'll need to pull it out in a future PR. I'm assuming Alert service is considered a child service of Folder service and so it will eventually get its own Folder service registry implementation and some kind of equivalent to `DeleteDashboardsInFolder` (so `DeleteAlertsInFolder`). 

I hope that makes sense. There's a bit more context in the description bullet point "Deletions related to dashboard deletions..."

suntala committed:
Fix test
